### ISSUE TITLE
Refactor API

### ISSUE THREAD

**Issue by m-pilia** [m-pilia/vim-yggdrasil#2]
The current procedural API was developed for internal use within [vim-ccls](https://github.com/m-pilia/vim-ccls), and exposes too much of the internal implementation, being prone to coupling with modules that use it.

Refactor the API using a data provider design pattern, analogous to the [TreeDataProvider](https://code.visualstudio.com/api/references/vscode-api#TreeDataProvider) of Visual Studio Code. The API should expose one function to create a tree buffer, accepting a data provider as argument to populate the tree.

**Comment by prabirshrestha** [comment]
@m-pilia Any timeline for this. Would like to start using this in vim-lsp

**Comment by m-pilia** [comment]
@prabirshrestha Sorry, this summer I had some big things to take care of, and I had to pause my side projects. I was planning to resume this and possibly getting it done in September.

**Comment by m-pilia** [comment]
Possible public API: TreeDataProvider (dict) + TreeItem (dict)

### TreeDataProvider

Vim event handling unsuitable for this purpose. Replace events with update method? (Possible alternatitves: observer pattern, update on some vim event).

```
TreeDataProvider = {
    'getChildren': {element -> ProviderResult(element)}, " return a list of elements
    'getParent':   {element -> ProviderResult(element)}, " optional method, return an element
    'getTreeItem': {element -> TreeItem(element)},       " return a TreeItem representing the element
}
```

### TreeItem

Fields similar to VSCode. Most of them not directly used, kept for future use. Let them be available to the user within the command function (dict function).

```
TreeItem = {
    'collapsibleState': v:true,     " mandatory; for rendering purpose
    'command': function(...),       " mandatory; command executed on user action (dict function)
    'contextValue': 'string',       " optional; reserved for future use/can be used within command
    'description': 'string',        " optional; reserved for future use/can be used within command
    'iconPath': 'string',           " optional; reserved for future use/can be used within command
    'id': 'string',                 " mandatory; identify the item in the tree
    'label': 'string',              " mandatory; to display in the tree view
    'resourceUri': 'string',        " optional; can be used within command
    'tooltip': 'string',            " optional; reserved for future use/can be used within command
}
```

### Expected usage

```viml

function! Children(root, ...) abort
    return a:0 ? a:1.children : a:root.children
endfunction

function! Command(x) dict abort
    echo a:x.something . ' is represented by node ' . l:self.label
endfunction

" User defines provider, minimal example based on some tree object `x`
root = {...}
provider = {
\ 'getChildren': function('Children', [root]),
\ 'getParent': {x -> x.parent},
\ 'getTreeItem': {x -> {
\    'collapsibleState': 'collapsed',
\    'command': function('Command', [x]),
\    'id': x.id,
\    'label': x.label,
\    },
\ }

tree = yggdrasil#tree#new(provider, ...)

tree.update()
```

**Comment by prabirshrestha** [comment]
Adding other vim-lsp collaborators for their thoughts on the api @mattn @thomasfaingnaert. FYI this would allow us to support tree for call hierarchy and other information in vim-lsp.

@m-pilia Can we also design the api such a way that it works for async data providers? Need to have some `loading`, `error` and `done` state. In `vim-fz` I had even proposed streaming api so as results come in it can be added with `fz#append` api https://github.com/mattn/vim-fz/issues/4 

If I want to hook a custom keybinding such as `t` for open in new tab` and `o` for open in current window how would I implement it?

**Comment by m-pilia** [comment]
> @m-pilia Can we also design the api such a way that it works for async data providers? Need to have some `loading`, `error` and `done` state. In `vim-fz` I had even proposed streaming api so as results come in it can be added with `fz#append` api [mattn/vim-fz#4](https://github.com/mattn/vim-fz/issues/4)

In the current version asynchronicity is accomplished with (optional) callbacks. I am not aware of any way to run VimL functions asynchronously, so one option would be to keep using callbacks (i.e. in the data provider), diverging from the VSCode API.

```
TreeDataProvider = {
    'getChildren': {callback[, element] -> ...},
    'getParent':   {callback, element -> ...},
    'getTreeItem': {callback, element -> ...},
}

" This is not implemented by the user, just called within the methods
function! ExampleCallback(status, ...) abort
    if a:status #== 'success'
        " do something with a:0
    endif
endfunction
```

Each method accepts as arguments one element to be queried, and one callback that will take the result as argument. Now the implementation of the provider can for instance launch async jobs.

```viml
" Minimal dummy async example
function! Children(callback, ...) abort
    call job_start([...], {'out_cb': {out -> a:callback('success', out)}})
endfunction

" The user defines a provider, minimal example based on some tree object `x`
provider = {
\ 'getChildren': function('Children', [callback]),
\ 'getParent': {callback, x -> callback(x.parent)},
\ 'getTreeItem': {callback, x -> callback({
\    'collapsibleState': 'collapsed',
\    'command': function('...'),
\    'id': x.id,
\    'label': x.label,
\    }),
\ }
```

I like the idea of a generic interface with minimal coupling, but I am a bit worried. The VSCode API is nice, but VimL has different features than TypeScript (mostly in a negative sense), so I am not sure everything still makes 100% sense in this context or if more rethinking is needed.

> If I want to hook a custom keybinding such as `t` for open in new tab`and`o` for open in current window how would I implement it?

One solution I would like is to not create any window or tab, and let the tree constructor `yggdrasil#tree#new(...)` open a new tree buffer in the current window. It would be responsibility of the user to create a window (or possibly tab) and then call the constructor there. This way the user has full control and flexibility, and the API stays minimal.

**Comment by m-pilia** [comment]
@prabirshrestha I started implementing some proof of concept for the refactoring in #7 (there is an example of usage in the README.md).

**Comment by m-pilia** [comment]
@prabirshrestha I have implemented a fix for this and for #5 (PR #7 and #8). If you think the API is ok for your use case, I can merge and close the issues. Suggestions for improvements are welcome of course.
